fix/stuckdoors.lua

```diff
@@ -0,0 +1,120 @@
+-- Fix doors that are frozen in 'open' state.
+-- This may happen after people mess with the game by (incorrectly) teleporting units or items
+-- A door may stick open if the map occupancy flags are wrong.
+
+--[====[
+
+    fix/stuckdoors
+    ===============
+    Fix doors that are stuck open due to incorrect map occupancy flags, eg due to
+    incorrect use of `teleport`.
+
+]====]
+
+if #{...} > 0 then
+    qerror('This script takes no arguments.')
+end
+
+-- Util function: use to filter down to only an active unit
+function activeUnitFilter(unit)
+    return dfhack.units.isActive(unit)
+end
+
+-- Util function: find out if there are any units on the tile with coordinates x,y,z
+function unitOnTile(x, y, z)
+    local units = dfhack.units.getUnitsInBox(x,y,z,x,y,z,activeUnitFilter)
+    return #(units) > 0
+end
+
+-- Util function: find out if there are any items on the tile with coordinates x,y,z
+-- There should be one item on the tile: the door itself.
+-- This method is not especially precise as it's doing a simple count check.
+-- A better one would be somehow linking the door building with the door item
```

Comment:
this is not actually that hard to do. `dfhack.buildings.findAtTile(pos)` will get you the door (you can check the type to make sure), and there's a `contained_items` vector that should get you a reference to the door item.

Reply:
contained_items are only for items *inside* the building. this includes the building materials and any "clutter", like with workshops or trade depots. I was just suggesting a way to identify the door item that was used to build the door. Your existing method of finding all items on the tile is a good one.

Reply:
There isn't a function like that that I know of. Is performance a problem? how long does the script take to run on a mature fort with lots of items and doors? This fort might be a good test case: https://dffd.bay12games.com/file.php?id=15434

Reply:
Yeah, 4-5 seconds is in the danger zone. A little performance optimization would be a good idea.

(Thank you! I've been working on that layout for 10 years : p.  I wrote quickfort so I could have better tools to iterate on the design)